src/sqlfluff/utils/reflow/elements.py

```diff
@@ -0,0 +1,331 @@
+"""Dataclasses for reflow work."""
+
+
+from itertools import chain
+import logging
+from dataclasses import dataclass
+from typing import List, Optional, Sequence, Set, Tuple, cast
+
+from sqlfluff.core.parser import BaseSegment, RawSegment
+from sqlfluff.core.parser.segments.raw import WhitespaceSegment
+from sqlfluff.core.rules.base import LintFix
+
+from sqlfluff.utils.reflow.config import ReflowConfig
+
+# We're in the utils module, but users will expect reflow
+# logs to appear in the context of rules. Hence it's a subset
+# of the rules logger.
+reflow_logger = logging.getLogger("sqlfluff.rules.reflow")
+
+
+@dataclass(frozen=True)
+class ReflowElement:
+    """Base reflow element class."""
+
+    segments: Sequence[RawSegment]
+
+    @staticmethod
+    def _class_types(segments: Sequence[RawSegment]) -> Set[str]:
+        return set(chain.from_iterable(seg.class_types for seg in segments))
+
+    @property
+    def class_types(self):
+        """The set of contained class types.
+
+        Parallel to BaseSegment.class_types
+        """
+        return self._class_types(self.segments)
+
+    @classmethod
+    def from_config(cls, segments, config: ReflowConfig):
+        """Extendable constructor which accepts config."""
+        return cls(segments=segments)
+
+
+@dataclass(frozen=True)
+class ReflowBlock(ReflowElement):
+    """Class for keeping track of elements to reflow.
+
+    It holds segments to reflow and also exposes configuration
+    around how they are expected to reflow around others.
+
+    The attributes exposed are designed to be "post configuration"
+    i.e. they should reflect configuration appropriately.
+
+    NOTE: These are the smallest unit of "work" within
+    the reflow methods, and may contain meta segments.
+    """
+
+    # Options for spacing rules are:
+    # - single: the default (one single space)
+    # - close: no whitespace
```

Comment:
None or false would be much more explicit, let's stick to one of those?

Reply:
ah - there are going to be more than three unfortunately I think. If I can keep the number down I agree, but that might not be possible.

I'm working on how to handle comments effectively now and that might make things clearer.